KeePass: add plugin support; include keefox

anicole opened the pull request:
Hi, I started playing with NixOS recently and I've tried to add keefox (http://keefox.org/) to nixpkgs. After multiple attempts I still worry if my approach is acceptable, and after reading contributing notes in the manual I still worry whether I'm requesting a pull correctly, but hopefully I won't cause too much trouble :)
#### Why this is so complicated:

Keefox has two parts: a browser add-on and a KeePass plugin.

KeePass looks for plugins in directory containing KeePass.exe; AFAIU, on NixOS it meant effectively no plugin support. The first approach I had working was to modify keepass derivation to copy plugin to that directory if configuration flag was enabled - this worked, but coupled app to plugin tightly in repository.

I discovered buildEnv, but KeePass.exe would need to copied to the resulting environment rather than linked for the correct directory to be searched. Additionally, compiler is required at runtime for the plugin to be compiled during loading. Most probably this could have been overcome with more complex layering and specific copying. I wanted to avoid patching keepass (would addition of environmental variable for search path be a potential security issue?), but in the end, that's what is done - paths to plugins are hardcoded into the nix store. I hope the result is cleaner.

The patch is inline and makes whitespace significant. I could not work out how to move patch to files and not get errors :(
#### Merging

I use 15.09 stable nixos channel, and I **have tested** my changes by building keepass with and without keefox using nix-env, on user account, without mono in path. The pull request is against master, as stated in manual; I **haven't** tested nixpkgs master. I wonder if I should do a pull against release/15.09 instead?

anicole committed:
Keefox (keepass plugin): init at 1.5.4

anicole committed:
keepass: load plugins from store paths

garbas commented:
@anicole is there a reason why not patch existing version of keepass?

also it would be really nice to have all patches in keepass.patch file and not inline. makes whole expression hard to read.

anicole commented:
@garbas Do you mean, why did I split keepass and keepass-with-plugins? No technical reasons. I thought it may be preferred to keep existing package separate for clarity and for people who'd prefer not to patch source. I may merge them, if it is preferred.

About moving patch to a file: yes, It'd be nicer. I've tried moving header, line and footer parts into files, read them in, do substitution in string from the line part, and concatenate all parts together. Loading string from files caused nix to raise error about hashing derivation (AFAIU - I did not understand fully). I will try again then. :)

garbas commented:
@anicole yes please merge them. patch doesn't really change anything from functionality of keepass, just makes it more suitable for nix

could you make default value of ``plugins`` argument having default value ``[]``? eg. ``plugins ? []`` 
it makes sense to have default empty list of plugins

please make that patch being in a file. if you need help paste traceback here and i'll help you.

anicole commented:
@garbas I've merged changes into keepass'es default.nix and moved patch into two fragment files. Also, I've updated keefox .xpi version to `1.5.4b2`.

`unsafeDiscardStringContext` helped me silence errors (`the string... is not allowed to refer to a store path`). If you have a moment, could you please point me to an explanation for what this error really means and especially why `unsafeDiscardStringContext` *has not* been needed with the patch inlined? Is it some technical detail of handling string concatenation with `+` vs only replace/concat, or is this derivation really unsafe?

Full stack:

    error: while evaluating the attribute ‘pluginLoadPathsPatch’ of the derivation ‘keepass-2.30’ at /git/nixpkgs/pkgs/build-support/build-dotnet-package/default.nix:17:7:
    while evaluating anonymous function at /git/nixpkgs/pkgs/applications/misc/keepass/default.nix:31:16, called from undefined position:
    the string ‘/nix/store/1qcj4pmraqp9zcdk9zhyf77v7x53gx5p-keefox-1.5.4b2’ is not allowed to refer to a store path (such as ‘!out!/nix/store/dcinbgf7dvkhccm88ddm0vnrma4gxrb3-keefox-1.5.4b2.drv’), at /git/nixpkgs/pkgs/applications/misc/keepass/default.nix:31:19
